## SciPy Tutorial

SciPy is a Python library which is used to solve scientific and mathematical problems

### single intergration

In [1]:
from scipy.integrate import quad  # allows us to integrate from limit a to b

In [2]:
# defining function for integration
def integratefunc(x):
    return x

In [3]:
quad(integratefunc,0,1)

(0.5, 5.551115123125783e-15)

here 5.55.. is the numerical error in computation

##### example 2

In [4]:
def integratefunc2(x,a,b):
    return x*a+b

a=3
b=2

In [5]:
quad(integratefunc2,0,1,args=(a,b))

(3.5, 3.885780586188048e-14)

##### example 3

In [6]:
from scipy import integrate 
func = lambda x: x**2 + x + 1
ans = integrate.quad(func, 1, 4) 
print(ans)
print(ans[0]) 

(31.5, 3.497202527569243e-13)
31.5


### double integration

In [7]:
import scipy.integrate as integrate

In [8]:
def f(x,y):
    return x+y
integrate.dblquad(f,0,1,lambda x:0, lambda x:2)
# range of x is 0,1 and y is 0,2

(3.0, 4.436070580899685e-14)

### optimization
* Optimization is a mathematical procedure to select the best decision variables from a set of alternatives to get the optimum objective value.
* used to improve performance of system mathematically by fine tuning process paameters

In [9]:
import numpy as np
from scipy import optimize

#### finding minimum value

In [10]:
def f(x):
    return x**2+5*np.sin(x)

In [11]:
minimal_value = optimize.minimize(f,x0=0,method='bfgs',options={'disp':True})  

Optimization terminated successfully.
         Current function value: -3.246394
         Iterations: 5
         Function evaluations: 12
         Gradient evaluations: 6


In [12]:
minimalValue=optimize.minimize(f,x0=0,method='bfgs') 
minimalValue

  message: Optimization terminated successfully.
  success: True
   status: 0
      fun: -3.2463942726915387
        x: [-1.111e+00]
      nit: 5
      jac: [-2.980e-08]
 hess_inv: [[ 1.544e-01]]
     nfev: 12
     njev: 6

#### finding roots

In [13]:
from scipy.optimize import root

In [14]:
def rootfunc(x):
    return x + 3.5 * np.cos(x)

In [15]:
root=root(rootfunc,0.3)  #0.3 is intial guess for root
root

 message: The solution converged.
 success: True
  status: 1
     fun: [ 0.000e+00]
       x: [-1.216e+00]
  method: hybr
    nfev: 14
    fjac: [[-1.000e+00]]
       r: [-4.282e+00]
     qtf: [-8.329e-13]

here x:[-1.21..] is the root

### matrix

In [16]:
from scipy import linalg

In [17]:
matrix=np.array([[1,2],[3,8]])
matrix

array([[1, 2],
       [3, 8]])

inversing matrix

In [18]:
linalg.inv(matrix)

array([[ 4. , -1. ],
       [-1.5,  0.5]])

determinant of matrix

In [19]:
linalg.det(matrix)

2.0

### solving linear functions

linear equations
* $ 2x+3y+z=21 $
* $-x + 5y +4z=9$
* $3x+2y+9z=6$

In [20]:
numarr=np.array([[2,3,1],[-1,5,4],[3,2,9]])
num_val=np.array([21,9,6])

In [21]:
linalg.solve(numarr,num_val)

array([ 4.95,  4.35, -1.95])

another method

In [25]:
from sympy import symbols, Eq, solve 
x,y,z=symbols('x,y,z')
eq1=Eq((2*x+3*y+z),21)
eq2=Eq((-1*x+5*y+4*z),9)
eq3=Eq((3*x+2*y+9*z),6)
solve((eq1,eq2,eq3),(x,y,z))

{x: 99/20, y: 87/20, z: -39/20}

### single value decomposition
* Singular Value Decomposition is a factorization of a matrix into three matrices

In [26]:
matrix=np.array([[1,2],[3,8]])
matrix.shape

(2, 2)

In [27]:
linalg.svd(matrix)

(array([[-0.25204813, -0.9677147 ],
        [-0.9677147 ,  0.25204813]]),
 array([8.8288552 , 0.22652994]),
 array([[-0.35737275, -0.93396184],
        [-0.93396184,  0.35737275]]))

U: Contains the left singular vectors


Σ: Diagonal matrix with singular values 

VH: Contains the right singular vectors

### Eigenvalues and Eigenvectors

In [30]:
# test_data matrix
test_data = np.array([[5,8],[7,9]])
eigenValues, eigenVector = linalg.eig(test_data)

In [31]:
# eignen values
eigenValues

array([-0.74596669+0.j, 14.74596669+0.j])

In [32]:
#eigenvectors
eigenVector

array([[-0.81220939, -0.63447346],
       [ 0.58336601, -0.77294465]])

### statistics

In [34]:
from scipy.stats import norm #normal distribution

##### generating 10 random numbers sampled from a normal distribution with mean 0 and standard deviation 1

In [35]:
norm.rvs(loc=0,scale=1,size=10)  #loc - mean, scale- standard deviation

array([-1.08747222, -0.28168722, -1.20237703, -0.21091911,  2.79678708,
       -0.02121145,  0.03691377, -0.14314732, -0.04800627, -1.74142768])

##### cumulative distribution function
probability that a random variable from this normal distribution is less than or equal to 5

In [36]:
norm.cdf(5,loc=1,scale=2) 

0.9772498680518208

##### probability density function
value of the probability density function of the normal distribution at point 9.

In [37]:
norm.pdf(9,loc=0,scale=1)

1.0279773571668917e-18

##### skewness and kurtosis 
* skewness- measure of symetry
* kurtosis- A measure of whether the data is heavy or lightly tailed

In [38]:
from scipy.stats import skew, kurtosis
v = np.random.normal(size=100)
print(skew(v))
print(kurtosis(v))

-0.22783577411826295
0.014464412953754557


##### statistical description

In [39]:
from scipy.stats import describe
v2 = np.random.normal(size=100)
res = describe(v2)
print(res)

DescribeResult(nobs=100, minmax=(-2.283443073637383, 2.4127592112524026), mean=0.06157167999360245, variance=1.001462835666611, skewness=-0.20303844840985388, kurtosis=-0.3807299345159221)


#### examples

Test has 30 questions worth 150 marks

Two types of questions will be there:
* True,False worth 4 marks each
* Multiple choice worth 9 points each

Let x be the no of true/false questions

Let y be the MCQ questions 

x + y = 30

4x + 9y = 150

In [42]:
num=np.array([[1,1],[4,9]])
sol=np.array([30,150])
linalg.solve(num,sol)

array([24.,  6.])

Value of x = 24, Value of y = 6

Means, True/False questions will be 24 & MCQ questions will be 6 in number